src/amici.cpp

```diff
@@ -1383,6 +1384,375 @@ int getDiagnosisB(int it, void *ami_mem, UserData *udata, ReturnData *rdata, Tem
     return status;
 }
 
+/* ------------------------------------------------------------------------------------- */
+/* ------------------------------------------------------------------------------------- */
+/* ------------------------------------------------------------------------------------- */
+
+
+int applyNewtonsMethod(void *ami_mem, UserData *udata, ReturnData *rdata, TempData *tdata, int status, int ntry) {
+    /**
+     * applyNewtonsMethod applies Newtons method to the current state x to find the steady state
+     *
+     * @param[in] ami_mem pointer to the solver memory block @type *void
+     * @param[in] udata pointer to the user data struct @type UserData
+     * @param[in] tdata pointer to the temporary data struct @type TempData
+     * @param[out] tdata pointer to the temporary data struct @type TempData
+     * @param[out] rdata pointer to the return data struct @type ReturnData
+     * @return void
+     */
+
+    int ix;
+    int inewt = 0;
+    int stat1 = -1;
+    
+    double res;
+    double rel_res;
+    double res_old;
+    double gamma = 1.0;
+    realtype *x_tmp;
+    
+    N_Vector delta = N_VNew_Serial(udata->nx);
+    N_Vector rel_x = N_VNew_Serial(udata->nx);
+    N_Vector tmp_x = N_VNew_Serial(udata->nx);
+    
+    N_VConst(0.0, delta);
+    
+    // Check, how fxdot is used exactly within AMICI...
+    fxdot(tdata->t, tdata->x, tdata->dx, tdata->xdot, udata);
+    res = sqrt(N_VDotProd(tdata->xdot,tdata->xdot));
+    
+    // Check for relative error, but make sure not to divide by 0!
+    // Ensure positivity of the state
+    N_VScale(1.0, tdata->x, tmp_x);
+    N_VAbs(tmp_x, tmp_x);
+    x_tmp = N_VGetArrayPointer(tmp_x);
```

Comment:
having `tmp_x` and `x_tmp` in the same function with different types is very confusing to me. I recommend renaming `tmp_x` to `x` or `x_newton`